File: plugins/dartai/scripts/track_subagent_spawn.py
#!/usr/bin/env python3
"""
Track Subagent Spawn - PostToolUse hook for Task tool.

This hook fires when a new subagent is spawned via the Task tool.
Tracks spawns for context isolation verification.
"""

import json
import os
from datetime import datetime
from pathlib import Path

STATE_DIR = Path.home() / ".dartai"
LOOP_FILE = Path(".claude/dartai-loop-state.json")


def track_spawn():
    """Track subagent spawn for context isolation verification."""
    # Ensure state directory exists
    STATE_DIR.mkdir(parents=True, exist_ok=True)

    # Load current loop state
    loop = {"iterations": 0, "spawns": 0, "started_at": None}
    if LOOP_FILE.exists():
        try:
            with open(LOOP_FILE) as f:
                loop = json.load(f)
        except Exception:
            pass

    # Increment spawn count
    loop["spawns"] = loop.get("spawns", 0) + 1
    loop["last_spawn_at"] = datetime.now().isoformat()

    # Get subagent type if available
    subagent_type = os.environ.get("CLAUDE_SUBAGENT_TYPE", "unknown")

    # Track spawn in log for audit trail
    spawn_log = STATE_DIR / "spawns.log"
    with open(spawn_log, "a") as f:
        f.write(f"{datetime.now().isoformat()} - Spawn #{loop['spawns']}: {subagent_type}\n")

    # Update loop file
    LOOP_FILE.parent.mkdir(parents=True, exist_ok=True)
    with open(LOOP_FILE, "w") as f:
        json.dump(loop, f, indent=2)

    return {
        "success": True,
        "spawn_count": loop["spawns"],
        "subagent_type": subagent_type
    }


if __name__ == "__main__":
    print(json.dumps(track_spawn()))

File: plugins/workflow/scripts/track-subagent-spawn.js
#!/usr/bin/env node
/**
 * Track when subagents are spawned (PostToolUse for Task tool)
 */

const fs = require('fs');
const path = require('path');

const workflowDir = path.join(process.cwd(), '.claude');
const sessionFile = path.join(workflowDir, 'workflow-session.json');

// Update spawn count
if (fs.existsSync(sessionFile)) {
  try {
    const sessionData = JSON.parse(fs.readFileSync(sessionFile, 'utf-8'));
    sessionData.subagent_spawns = (sessionData.subagent_spawns || 0) + 1;
    fs.writeFileSync(sessionFile, JSON.stringify(sessionData, null, 2));
  } catch {
    // Ignore errors
  }
}

// Log spawn (helps track context isolation)
const logPath = path.join(workflowDir, 'workflow-spawns.log');
try {
  fs.appendFileSync(logPath, `Subagent spawned at ${new Date().toISOString()}\n`);
} catch {
  // Ignore errors
}

process.exit(0);
